### PR TITLE
SPARK-5112. Expose SizeEstimator as a developer api

### PR DESCRIPTION
**Pull request by sryza**
"The best way to size the amount of memory consumption your dataset will require is to create an RDD, put it into cache, and look at the SparkContext logs on your driver program. The logs will tell you how much memory each partition is consuming, which you can aggregate to get the total size of the RDD."
-the Tuning Spark page

This is a pain. It would be much nicer to expose simply functionality for understanding the memory footprint of a Java object.

### CODE REVIEW

**Comment by pwendell** [comment]
That doc is very outdated - you can actually just look in the UI after caching some data, you don't need to visit the logs.

**Comment by sryza** [comment]
Ooh OK I'll update the doc.  That's still a little cumbersome though for someone who just wants to see how much space an object takes up.  Most of the recommendations on that page are at the micro level - tuning the memory taken up by a single object.  It would be useful to have a way to determine this amount directly.

E.g. if I have plan to have an RDD of some case class I might want to see how much space each instance takes up.  Then I can experiment quickly with things like flatter structures or removing fields and see if they improve the footprint.

Another thing is that users trying to tune applications commonly need to figure out whether they should broadcast side-data or figure out a way to load it as an RDD and join.  It would be helpful to be able to have a single function call that's easy to use from the shell to find out the size of the data you're about to broadcast.

**Comment by pwendell** [comment]
Just not sure how overall useful it would be. For RDD data, it might be slightly misleading here because of things like serialization in-memory. For broadcast objects in the shell, it would only work in the scala shell though because of the way that serialization works in Python.

I'm also not totally sure overall how accurate our memory estimation is and it may get less so if we add smarter caching for SchemaRDD's. Anyways, what would be helpful, could you walk through an example with a case class or something and show how accurate it is? That would be useful to better understand.

One thing we could do that would be more isolated is have a function in SparkContext called `estimateSizeOf(object: Any)`, so that at least we don't expose the class location and names as API's.

**Comment by srowen** [comment]
Should this just turn into a doc update PR then?

**Comment by sryza** [comment]
Adding an `estimateSizeOf` method to `SparkContext` sounds reasonable to me.

I agree that there's not a great way to expose something like this for Python. But I don't think the zaniness of Python-JVM interaction means that we shouldn't expose useful functionality to pure-JVM apps.

> For RDD data, it might be slightly misleading here because of things like serialization in-memory.

I think this is the kind of thing we can just document.  Adding a separate `estimateSerializedSizeOf` method would be helpful as well.

> I'm also not totally sure overall how accurate our memory estimation is and it may get less so if we add smarter caching for SchemaRDD's.

I've found it to be very accurate in my experiments.  We rely on its accuracy for shuffle memory management and POJO caching, so to the extent that it's inaccurate we've got bigger problems.

**Comment by shivaram** [comment]
The size estimator should be pretty accurate for measuring the size of small Java objects. You could add a note that we do some approximation for large arrays (we sample some elements of the array and then extrapolate).

**Comment by srowen** [comment]
I'd like to resolve this one way or the other. My hesitation is mostly about tacking on another method to `SparkContext`, developer API or no. Would it really be a better way to estimate actual RDD size, since that depends on other stuff (even the JVM, whether data is serialized, etc.)

**Comment by squito** [comment]
I think this is a good change.  Yes, you could cache an RDD and see its size, but think about what a pain that actually is if you wanted to do it programmatically.  You'd need to register a spark listener, wait to get the appropriate events and look at the sizes.

If it was easy to do this programmatically via an RDD, then I'd say this change isn't necessary.  Eg., if you could do something like

```
val (_, meta) = sc.parallelize(oneObject).cache().count()
meta.getPartitionSizes
```

Then we wouldn't need to expose this.  But that's an even bigger api change, and one that I would be far more nervous about (the code above is definitely not a viable alternative, lots of reasons it doesn't make sense).

File: core/src/main/scala/org/apache/spark/util/SizeEstimator.scala [hunk 1]
```diff
@@ -28,6 +28,7 @@ import java.util.concurrent.ConcurrentHashMap
 import scala.collection.mutable.ArrayBuffer
 
 import org.apache.spark.Logging
+import org.apache.spark.annotation.DeveloperApi
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/util/SizeEstimator.scala]: nit: this import isn't necessary anymore

**Comment by sryza** [comment]
Again, one of the main uses is estimating the size of variables you're considering broadcasting.  Another is experimenting with different representations - e.g. how much more efficient is declaring a custom class than just using a hash map?  In these situations, putting the data into and RDD to estimate the size would be an inconvenience.

**Comment by srowen** [comment]
OK, I think my remaining comment is just: why route through `SparkContext`? can this method be exposed where it is? It doesn't seem like it has to be tied to the context.

**Comment by sryza** [comment]
@pwendell thought this would be preferable:

> One thing we could do that would be more isolated is have a function in SparkContext called estimateSizeOf(object: Any), so that at least we don't expose the class location and names as API'

**Comment by pwendell** [comment]
The reason I proposed to put in SparkContext is to avoid committing to the current namespace/package of that object and just expose a narrower utility function off of SparkContext. Overall our estimation code is likely to evolve in the future. 

In terms of exposing or not, I'm okay to expose it given the reasons here. But can we give some warning to set expectations for the user? This estimation can be really inaccurate because of sampling and heuristics used internally. This is especially true if you have, say, a hashmap that has skewed keys - it will only sample a small percentage of all the keyspace and could miss hot keys.

So I'd just say it's an estimate of the in-memory size and uses sampling internally for complex objects. I think this is also the gist of @shivaram's suggestion.

**Comment by srowen** [comment]
OK how about just moving it up a level out of `util`, if that's generally not public, and making `SizeEstimator` public but a developer API? You have to commit to the method living somewhere and I think that's no more complex than routing it through the context API.

**Comment by pwendell** [comment]
Sure, sounds fine to me. We can have a static method for it.

**Comment by srowen** [comment]
Looking good though this needs a rebase now.

**Comment by pwendell** [comment]
Pending an update LGTM

**Comment by pwendell** [comment]
Hey @sryza and @srowen, sorry to vacillate here, but after looking at this more I really do think it would be better to just make this a static method on `SparkContext`. The reason is the following:

1. For users, it's much nicer to just call `SparkContext.estimateSize()` since in may cases they've already imported SparkContext. Having them import some other class for this one static function seems a little strange.
2. For developers, we now have this whole class exposed and a lot more room to accidentally expose new functionally if this class is later changed. There is a single static function buried deep in the middle of the class description that is public.

Basically I don't see a big motivation for having a separate class for this, and it forces us to make some longer term commitment to that class name, having static utility classes in the top level namespace, etc.

**Comment by srowen** [comment]
Hm, my gut is that it's stranger to put this in the context when it has nothing to do with the context. It's all private except for the one method. Lots of accidents can happen but, y'know, that's an argument anywhere. 

Something has to commit to at least one name somewhere, yes. What about a single `object` at this level that exposes a single public method, that calls into the existing class where it is? I suppose that minimizes the amount that is surfaced and keeps the implementation alongside other package-private stuff, at the cost of a little redundancy and extra declaration.

Now it really is down to where to place a single method, and I'd still argue a little for not tacking it on the context as an uber utility class. This is a developer API so convenience for the end user isn't a big deal.

**Comment by pwendell** [comment]
I'm good with the new idea you proposed. We keep `util.SizeEstimator` as is, but we have a public class in the root namespace `SizeEstimator` that just has a single method that calls into the internal one. We can always merge them together later if we want too, but IMO it's worth keeping the larger implementation totally private.

**Comment by srowen** [comment]
@sryza do you want to try to update this and get it in for 1.4? might still technically be time

**Comment by sryza** [comment]
Updated with the suggested approach.  Let me know what you think.

**Comment by srowen** [comment]
LGTM. I think the test failure was unrelated; let's see.

**Comment by srowen** [comment]
Jenkins retest this please.

**Comment by rxin** [comment]
Actually I strongly oppose putting this in the top level package. We will end up with a lot of random util objects or top level classes at the top.

One way to do this is to have a public util package, and move everything in util into internalutil if we want to hide utils.

**Comment by srowen** [comment]
@rxin yeah I hear you. It's not too late to move this. I struggle a bit with where to put this then. So you mean move `util` to `util.internal` or something, and move public utilities into `util`? makes sense to me. 

There are probably a few other things that could move down out of `org.apache.spark` that are not already public. Like the HTTP server stuff for example?
And I have thought and heard a few times that it might be nice to break up `Utils` into a few logical units.

It would be a big invasive change to the code; that's the only thing. I can take a swing at it if anyone voices support for that kind of thing?

**Comment by rxin** [comment]
I think it's a good idea to do that. For now why don't we move SizeEstimator back into util, rename the current util.SizeEstimator util.SizeEstimator0?

For 1.5 we can move stuff around into utils or utils.internal (seems long ... maybe we can find better names too).

**Comment by srowen** [comment]
I think @pwendell 's sentiment was that `util` shouldn't have public classes/objects -- what do you think Patrick? as an interim measure?

**Comment by rxin** [comment]
But we already have a bunch of stuff in util that's public ...
https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.util.package

**Comment by pwendell** [comment]
I'm okay to nest it under util, per Reynold's suggesting.

**Comment by srowen** [comment]
@sryza do you mind doing the honors as an part 2 for this issue?

**Comment by sryza** [comment]
@pwendell @rxin if we're putting it in the same package, I don't really understand why we don't just have a single class `SizeEstimator` with a single public method `estimate`?  We're worried in the future that someone might make other attributes of `SizeEstimator` public?  I don't see why we wouldn't trust ourselves to properly restrict access when a change proposing that comes around.  Or, in `SizeEstimator` / `SizeEstimator0` world, it wouldn't be just as easy to add something public to `SizeEstimator`.

**Comment by rxin** [comment]
I'm fine putting them in the same class too, provided that everything else is marked as private (I think it is already, isn't it?)

**Comment by sryza** [comment]
Is that ok with you as well @pwendell ?

**Comment by pwendell** [comment]
Okay - I officially give in and will not comment further on this thread :P
